src/main/java/uk/gov/cshr/controller/account/AgencyTokenVerificationController.java

```diff
@@ -0,0 +1,156 @@
+
+    private static final String CODE_ATTRIBUTE = "code";
+    private static final String VERIFY_TOKEN_FORM_TEMPLATE = "verifyTokenForm";
+    private static final String VERIFY_TOKEN_TEMPLATE = "verifyToken";
+    private static final String REDIRECT_VERIFY_TOKEN = "redirect:/account/verify/agency/";
+    private static final String REDIRECT_REACTIVATED_SUCCESS = "redirect:/account/reactivate/updated";
+    private static final String REDIRECT_ACCOUNT_EMAIL_UPDATED_SUCCESS = "redirect:/account/email/updated";
+
+    private static final String EMAIL_ATTRIBUTE = "email";
+
+    private final EmailUpdateService emailUpdateService;
+
+    private final CsrsService csrsService;
+
+    private final IdentityService identityService;
+
+    private final AgencyTokenCapacityService agencyTokenCapacityService;
+
+    private final ReactivationService reactivationService;
+
+    private final VerificationCodeDeterminationService verificationCodeDeterminationService;
+
+    public AgencyTokenVerificationController(
+            EmailUpdateService emailUpdateService,
+            CsrsService csrsService,
+            AgencyTokenCapacityService agencyTokenCapacityService,
+            IdentityService identityService,
+            VerificationCodeDeterminationService verificationCodeDeterminationService,
+            ReactivationService reactivationService) {
+        this.emailUpdateService = emailUpdateService;
+        this.csrsService = csrsService;
+        this.agencyTokenCapacityService = agencyTokenCapacityService;
+        this.identityService = identityService;
+        this.verificationCodeDeterminationService = verificationCodeDeterminationService;
+        this.reactivationService = reactivationService;
+    }
+
+    @GetMapping(path = "/{code}")
+    public String enterToken(Model model, @PathVariable String code) {
+        log.info("User accessing token-based verification screen");
+
+        if (model.containsAttribute(VERIFY_TOKEN_FORM_TEMPLATE)) {
+            addOrganisationsToModel(model);
+            model.addAttribute(CODE_ATTRIBUTE, code);
+            return VERIFY_TOKEN_TEMPLATE;
+        } else {
+            addOrganisationsToModel(model);
+
+            VerifyTokenForm form = new VerifyTokenForm();
+
+            model.addAttribute(VERIFY_TOKEN_FORM_TEMPLATE, form);
+            model.addAttribute(CODE_ATTRIBUTE, code);
+
+            return VERIFY_TOKEN_TEMPLATE;
+        }
+    }
+
+    @PostMapping(path = "/{code}")
+    public String checkToken(Model model,
+                             @ModelAttribute @Valid VerifyTokenForm form,
+                             BindingResult bindingResult,
+                             RedirectAttributes redirectAttributes) {
+        if (bindingResult.hasErrors()) {
+            buildGenericErrorModel(model, form);
+            return VERIFY_TOKEN_TEMPLATE;
+        }
+
+        try {
+            log.info("Token validation with values: {}", form.toString());
```

Comment:
Tiny optimisation when logging, when using logging templates (the {} bit), the logging engine will automatically to the toString call on the passed object, it will only do it if logging at the level called is enabled, this means the toString will not be called if it is not needed, in the current statement, toString will always be called even if the message is not logged.

Reply:
Ah good catch, I didn't know that 👍 

Removed `.toString()` from objects in logging calls.